fuel-types/src/canonical.rs

```diff
@@ -430,23 +430,40 @@ impl<const N: usize, T: Deserialize> Deserialize for [T; N] {
         } else {
             // Spec doesn't say how to deserialize arrays with unaligned
             // primitives(as `u16`, `u32`, `usize`), so unpad them.
-            let mut uninit = <MaybeUninit<[T; N]>>::uninit();
+            let mut uninit: [MaybeUninit<T>; N] =
+                [const { MaybeUninit::<T>::uninit() }; N];
             // The following line coerces the pointer to the array to a pointer
             // to the first array element which is equivalent.
             let mut ptr = uninit.as_mut_ptr() as *mut T;
-            for _ in 0..N {
-                let decoded = T::decode_static(buffer)?;
-                // SAFETY: We do not read uninitialized array contents
-                // 		 while initializing them.
-                unsafe {
-                    core::ptr::write(ptr, decoded);
+            for i in 0..N {
+                match T::decode_static(buffer) {
+                    Err(e) => {
+                        for item in uninit.iter_mut().take(i) {
+                            // SAFETY: all elements up to index i (excluded have been
+                            // initialised)
+                            unsafe {
+                                item.assume_init_drop();
+                            }
+                        }
+                        return Err(e);
+                    }
+                    Ok(decoded) => {
+                        // SAFETY: We do not read uninitialized array contents
+                        // 		 while initializing them.
+                        unsafe {
+                            core::ptr::write(ptr, decoded);
+                        }
+
+                        // SAFETY: Point to the next element after every iteration.
+                        // 		 We do this N times therefore this is safe.
+                        ptr = unsafe { ptr.add(1) };
+                    }
                 }
-                // SAFETY: Point to the next element after every iteration.
-                // 		 We do this N times therefore this is safe.
-                ptr = unsafe { ptr.add(1) };
             }
+            let final_ptr = &mut uninit as *mut _ as *mut [T; N];
             // SAFETY: All array elements have been initialized above.
-            let init = unsafe { uninit.assume_init() };
+            let init = unsafe { final_ptr.read() };
```

Comment:
This is ugly, but `mem::transmute` cannot be used for arrays with generic length. 

See https://github.com/rust-lang/rust/issues/61956
(The solution advocated there also mentions calling `mem::forget(uninit)`, but I don't think in this case is necessary and apparently it causes a compilation error.